<h1>Project 2 </h1>


perform logistic regression using the following data set

Use the feature sets given below to construct a logistic regression model 
 'ed', 'employ', 'equip',   'callcard', 'wireless'

CASE 1:- 

In [1]:
#start here 
#remember to use different code blocks for different segments 

import pandas as pd
import pylab as pl
import numpy as np
import scipy.optimize as opt
from sklearn import preprocessing
from IPython import display
import matplotlib.pyplot as plt

df=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%203/data/ChurnData.csv")
#display(df)

dff = df[["ed","employ","equip","callcard","wireless"]]
dff.head()

X = np.asanyarray(dff[["ed","employ","equip","callcard","wireless"]])

y = np.asanyarray(df["churn"])



from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)


from sklearn.model_selection import train_test_split
X_train, X_test , y_train , y_test = train_test_split(X,y,test_size = 0.2, random_state = 4)

print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C = 0.01 , solver = 'liblinear').fit(X_train,y_train)
LR

yhat = LR.predict(X_test)
yhat

yhat_prob = LR.predict_proba(X_test)
yhat_prob

a = confusion_matrix(y_test , yhat)
print(a)
print((a[0][0]+a[1][1])/(a[0][0]+a[0][1]+a[1][0]+a[1][1]))




# please remember to train and test your model according to the feature set selected 


Train set: (160, 5) (160,)
Test set: (40, 5) (40,)
[[18  7]
 [ 9  6]]
0.6


case 2 :- 

In [2]:
## now trying to get higher accuracy with different featured sets:-

import pandas as pd
import pylab as pl
import numpy as np
import scipy.optimize as opt
from sklearn import preprocessing
from IPython import display
import matplotlib.pyplot as plt

df=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%203/data/ChurnData.csv")


dff = df[["ed","employ","equip","logtoll","loglong"]]
dff.head()

X = np.asanyarray(dff[["ed","employ","equip","logtoll","loglong"]])

y = np.asanyarray(df["churn"])



from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)


from sklearn.model_selection import train_test_split
X_train, X_test , y_train , y_test = train_test_split(X,y,test_size = 0.2, random_state = 4)

print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C = 0.01 , solver = 'liblinear').fit(X_train,y_train)
LR

yhat = LR.predict(X_test)
yhat

yhat_prob = LR.predict_proba(X_test)
yhat_prob

a = confusion_matrix(y_test , yhat)
print(a)
print((a[0][0]+a[1][1])/(a[0][0]+a[0][1]+a[1][0]+a[1][1]))




Train set: (160, 5) (160,)
Test set: (40, 5) (40,)
[[23  2]
 [ 8  7]]
0.75


CASE 3 :- 

In [3]:
import pandas as pd
import pylab as pl
import numpy as np
import scipy.optimize as opt
from sklearn import preprocessing
from IPython import display
import matplotlib.pyplot as plt

df=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%203/data/ChurnData.csv")


dff = df[["age","employ","longmon","logtoll","loglong"]]
dff.head()

X = np.asanyarray(dff[["age","employ","longmon","logtoll","loglong"]])

y = np.asanyarray(df["churn"])



from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)


from sklearn.model_selection import train_test_split
X_train, X_test , y_train , y_test = train_test_split(X,y,test_size = 0.2, random_state = 4)

print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C = 0.01 , solver = 'liblinear').fit(X_train,y_train)
LR

yhat = LR.predict(X_test)
yhat

yhat_prob = LR.predict_proba(X_test)
yhat_prob

a = confusion_matrix(y_test , yhat)
print(a)
print((a[0][0]+a[1][1])/(a[0][0]+a[0][1]+a[1][0]+a[1][1]))



Train set: (160, 5) (160,)
Test set: (40, 5) (40,)
[[23  2]
 [ 7  8]]
0.775


CASE 4 :- 

In [4]:
import pandas as pd
import pylab as pl
import numpy as np
import scipy.optimize as opt
from sklearn import preprocessing
from IPython import display
import matplotlib.pyplot as plt

df=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%203/data/ChurnData.csv")


dff = df[["lninc","custcat","tenure","confer","ebill"]]
dff.head()

X = np.asanyarray(dff[["lninc","custcat","tenure","confer","ebill"]])

y = np.asanyarray(df["churn"])



from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)


from sklearn.model_selection import train_test_split
X_train, X_test , y_train , y_test = train_test_split(X,y,test_size = 0.2, random_state = 4)

print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C = 0.01 , solver = 'liblinear').fit(X_train,y_train)
LR

yhat = LR.predict(X_test)
yhat

yhat_prob = LR.predict_proba(X_test)
yhat_prob

a = confusion_matrix(y_test , yhat)
print(a)
print((a[0][0]+a[1][1])/(a[0][0]+a[0][1]+a[1][0]+a[1][1]))




Train set: (160, 5) (160,)
Test set: (40, 5) (40,)
[[22  3]
 [11  4]]
0.65


CASE 5 :- 

In [7]:
import pandas as pd
import pylab as pl
import numpy as np
import scipy.optimize as opt
from sklearn import preprocessing
from IPython import display
import matplotlib.pyplot as plt

df=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%203/data/ChurnData.csv")


dff = df[["callcard","income","internet","longmon","custcat"]]
dff.head()

X = np.asanyarray(dff[["callcard","income","internet","longmon","custcat"]])

y = np.asanyarray(df["churn"])



from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)


from sklearn.model_selection import train_test_split
X_train, X_test , y_train , y_test = train_test_split(X,y,test_size = 0.2, random_state = 4)

print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C = 0.01 , solver = 'liblinear').fit(X_train,y_train)
LR

yhat = LR.predict(X_test)
yhat

yhat_prob = LR.predict_proba(X_test)
yhat_prob

a = confusion_matrix(y_test , yhat)
print(a)
print((a[0][0]+a[1][1])/(a[0][0]+a[0][1]+a[1][0]+a[1][1]))




Train set: (160, 5) (160,)
Test set: (40, 5) (40,)
[[25  0]
 [11  4]]
0.725


## So i tested most of the cases and according to my predictions the most accuracy achieved is 77.5% when we consider age, employ , longmon , loglong and logtoll as our feature sets